forc-pkg/src/pkg.rs

```diff
@@ -2187,21 +2271,35 @@ pub fn build(plan: &BuildPlan, profile: &BuildProfile) -> anyhow::Result<(Compil
     let mut storage_slots = vec![];
     let mut bytecode = vec![];
     let mut tree_type = None;
+    let root_node = plan
+        .compilation_order()
+        .last()
+        .expect("Compilation order is empty!");
+    let root_node_pinned = &plan.graph()[*root_node];
+    let root_manifest = &plan.manifest_map()[&root_node_pinned.id()];
+    let mut compiled_map = HashMap::new();
     for &node in &plan.compilation_order {
         let pkg = &plan.graph()[node];
         let manifest = &plan.manifest_map()[&pkg.id()];
         let constants = manifest.config_time_constants();
-        let dep_namespace = match dependency_namespace(&namespace_map, &plan.graph, node, constants)
-        {
-            Ok(o) => o,
-            Err(errs) => {
-                print_on_failure(profile.terse, &[], &errs);
-                bail!("Failed to compile {}", pkg.name);
-            }
-        };
-
+        let dep_namespace =
+            match dependency_namespace(&namespace_map, &compiled_map, &plan.graph, node, constants)
+            {
+                Ok(o) => o,
+                Err(errs) => {
+                    print_on_failure(profile.terse, &[], &errs);
+                    bail!("Failed to compile {}", pkg.name);
+                }
+            };
         let res = compile(pkg, manifest, profile, dep_namespace, &mut source_map)?;
         let (compiled, maybe_namespace) = res;
+        // If the current node is a contract dependency, collect the contract_id
+        if root_manifest
+            .contract_deps()
+            .any(|dep| *dep.0 == manifest.project.name)
+        {
+            compiled_map.insert(node, compiled.clone());
+        }
```

Comment:
This still doesn't account for the case where the package is only a contract dependency of another contract dependency.

We can remove all the `root_`-specific logic and instead check if the current node is depended on as a contract dependency by any other node.

We can do so by checking if any of the incoming edges of this node are `DepKind::Contract`, e.g. something like:

```rust
if plan.graph().edges_directed(node, Incoming).any(|e| e.weight().kind == DepKind::Contract) {
    compiled_map.insert(node, compiled.clone());
}
```

Reply:
Ahh, yeah I missed this part thanks for pointing out! I Maybe I should add this case to the test too